File: projects/03-realtime-pubsub/src/metrics.rs
//! Prometheus metrics for the observability checklist (see SPEC.md → Horizontal
//! checklist → Observability).
//!
//! Same pattern as `01-url-shortener/src/metrics.rs`: the [`metrics`] facade
//! writes to a process-global recorder, so call sites elsewhere in the crate
//! (`hub.rs`, `backpressure.rs`, `routes.rs`, `presence.rs`) just name a metric
//! and stay decoupled from this wiring. Until [`install`] runs, the macros are
//! no-ops — exactly what tests want.
//!
//! [`install`] returns a [`PrometheusHandle`]; render it from `/metrics` (see
//! [`crate::routes::metrics_router`]) for a scrape endpoint.
//!
//! ## What's wired vs. what's a TODO
//! The recorder + `/metrics` endpoint are wired below. Drop counting on the
//! publish path lives in [`Mailbox::deliver`](crate::backpressure::Mailbox::deliver)
//! ([`MESSAGES_DROPPED_TOTAL`]). Remaining call sites are still **TODO**:
//! - `hub.rs::Hub::publish` — [`MESSAGES_DELIVERED_TOTAL`],
//!   [`SLOW_CLIENT_DISCONNECTS_TOTAL`] when a publish triggers disconnect cleanup.
//! - `routes.rs::dispatch` — increment [`MESSAGES_PUBLISHED_TOTAL`] when a client
//!   sends `Publish`.
//! - `routes.rs::handle_socket` — gauge [`OPEN_CONNECTIONS`] up on connect, down
//!   on teardown.
//! - `hub.rs` / subscribe paths — refresh [`SUBSCRIPTIONS_TOTAL`] and
//!   [`TOPICS_TOTAL`] gauges when membership changes (or sample from the hub map).
//! - `presence.rs` — gauge [`PRESENCE_MEMBERS`] per topic (or a single aggregate
//!   if you prefer fewer series).

use metrics_exporter_prometheus::{PrometheusBuilder, PrometheusHandle};

/// Client-originated publishes accepted by the server. Increment in
/// `routes::dispatch` on `ClientMessage::Publish`.
pub const MESSAGES_PUBLISHED_TOTAL: &str = "realtime_pubsub_messages_published_total";

/// Fan-out deliveries that reached a subscriber's bounded mailbox (`DeliverOutcome::Delivered`).
/// Increment in `hub::Hub::publish` per subscriber reached.
pub const MESSAGES_DELIVERED_TOTAL: &str = "realtime_pubsub_messages_delivered_total";

/// Messages shed by the backpressure policy — the whole point of V2. Increment
/// on `DeliverOutcome::Dropped` and on `DropOldest` evictions. Label `policy`
/// with [`OverflowPolicyLabel`](OverflowPolicyLabel) (`drop_newest`, `drop_oldest`,
/// `disconnect`).
pub const MESSAGES_DROPPED_TOTAL: &str = "realtime_pubsub_messages_dropped_total";

/// Connections torn down because a full outbox triggered the `disconnect` policy
/// (or the mailbox was closed). Increment when `publish` reaps a wedged subscriber.
pub const SLOW_CLIENT_DISCONNECTS_TOTAL: &str = "realtime_pubsub_slow_client_disconnects_total";

/// Live WebSocket connections. Increment in `handle_socket` on connect, decrement
/// on every exit path.
pub const OPEN_CONNECTIONS: &str = "realtime_pubsub_open_connections";

/// Total active topic subscriptions across all connections (sum of hub map sizes).
pub const SUBSCRIPTIONS_TOTAL: &str = "realtime_pubsub_subscriptions_total";

/// Distinct topics with at least one subscriber.
pub const TOPICS_TOTAL: &str = "realtime_pubsub_topics_total";

/// Current room membership count. Consider a `topic` label for per-room visibility;
/// watch cardinality if topic names are client-controlled.
pub const PRESENCE_MEMBERS: &str = "realtime_pubsub_presence_members";

/// Install the process-global Prometheus recorder and return a handle used to
/// render the registry for `/metrics`. Call once, from `main`, after telemetry
/// init. Panics if a recorder is already installed (calling it twice is a bug).
pub fn install() -> PrometheusHandle {
    register_descriptions();
    PrometheusBuilder::new()
        .install_recorder()
        .expect("install prometheus recorder")
}

fn register_descriptions() {
    let published = MESSAGES_PUBLISHED_TOTAL;
    let delivered = MESSAGES_DELIVERED_TOTAL;
    let dropped = MESSAGES_DROPPED_TOTAL;
    let slow_disconnects = SLOW_CLIENT_DISCONNECTS_TOTAL;
    let open_connections = OPEN_CONNECTIONS;
    let subscriptions = SUBSCRIPTIONS_TOTAL;
    let topics = TOPICS_TOTAL;
    let presence = PRESENCE_MEMBERS;

    metrics::describe_counter!(
        published,
        "Client-originated publishes accepted by the server"
    );
    metrics::describe_counter!(
        delivered,
        "Fan-out deliveries that reached a subscriber mailbox"
    );
    metrics::describe_counter!(
        dropped,
        "Messages shed by the backpressure overflow policy, labelled by policy"
    );
    metrics::describe_counter!(
        slow_disconnects,
        "Connections disconnected because the outbound mailbox was full or closed"
    );
    metrics::describe_gauge!(open_connections, "Live WebSocket connections");
    metrics::describe_gauge!(
        subscriptions,
        "Active topic subscriptions across all connections"
    );
    metrics::describe_gauge!(topics, "Distinct topics with at least one subscriber");
    metrics::describe_gauge!(
        presence,
        "Current room membership count, optionally labelled by topic"
    );
}

File: projects/03-realtime-pubsub/src/realtime_pubsub/metrics.py
"""Prometheus metrics for the observability checklist (see SPEC.md -> Horizontal
checklist -> Observability).

Python shape note: the Rust side used the `metrics` facade — string constants
written to a process-global recorder, installed once from `main`. `prometheus_client`
has no facade and needs none: a metric *is* an object, created once at import and
registered on the default `REGISTRY`. Call sites import the object and call
`.inc()` / `.set()` on it. That means these are live from the moment the module
is imported, including under pytest, so a test can assert on a counter without
installing anything first.

`common_telemetry.metrics_routes()` renders that same default registry at
`GET /metrics`, so nothing here has to be wired into the app by hand.

## What's wired vs. what's a TODO
Drop counting on the publish path is wired in `backpressure.Mailbox.deliver`
(`MESSAGES_DROPPED`), and `hub.Hub.publish` records deliveries and slow-client
reaps. Remaining call sites are still **TODO** — see the SPEC's Observability
checklist:
  - `routes.dispatch` — `MESSAGES_PUBLISHED` when a client sends `publish`.
  - `routes.websocket_endpoint` — `OPEN_CONNECTIONS` up on connect, down on
    every teardown path.
  - subscribe/unsubscribe paths — refresh `SUBSCRIPTIONS` and `TOPICS`.
  - `presence.py` — `PRESENCE_MEMBERS` per topic (or one aggregate if you would
    rather not pay the cardinality).
"""

from __future__ import annotations

from prometheus_client import Counter, Gauge

__all__ = [
    "MESSAGES_DELIVERED",
    "MESSAGES_DROPPED",
    "MESSAGES_PUBLISHED",
    "OPEN_CONNECTIONS",
    "PRESENCE_MEMBERS",
    "SLOW_CLIENT_DISCONNECTS",
    "SUBSCRIPTIONS",
    "TOPICS",
]

MESSAGES_PUBLISHED = Counter(
    "realtime_pubsub_messages_published_total",
    "Client-originated publishes accepted by the server",
)

MESSAGES_DELIVERED = Counter(
    "realtime_pubsub_messages_delivered_total",
    "Fan-out deliveries that reached a subscriber mailbox",
)

MESSAGES_DROPPED = Counter(
    "realtime_pubsub_messages_dropped_total",
    "Messages shed by the backpressure overflow policy, labelled by policy",
    labelnames=("policy",),
)
"""The whole point of V2. Labelled by `OverflowPolicy.value` so a dashboard can
tell a drop_oldest eviction from a drop_newest refusal."""

SLOW_CLIENT_DISCONNECTS = Counter(
    "realtime_pubsub_slow_client_disconnects_total",
    "Connections disconnected because the outbound mailbox was full or closed",
)

OPEN_CONNECTIONS = Gauge(
    "realtime_pubsub_open_connections",
    "Live WebSocket connections",
)

SUBSCRIPTIONS = Gauge(
    "realtime_pubsub_subscriptions_total",
    "Active topic subscriptions across all connections",
)

TOPICS = Gauge(
    "realtime_pubsub_topics_total",
    "Distinct topics with at least one subscriber",
)

PRESENCE_MEMBERS = Gauge(
    "realtime_pubsub_presence_members",
    "Current room membership count",
)
